packages/toolpad-core/src/types/Navigation.types.ts

```diff
@@ -0,0 +1,13 @@
+export interface NavigationRoute {
+  label: string;
+  path?: string;
+  icon: React.ReactNode;
+  routes?: (Omit<NavigationRoute, 'routes'> & { path: string })[];
+}
+
+export interface NavigationSection {
+  title: string;
+  routes: NavigationRoute[];
+}
+
+export type Navigation = NavigationSection[];
```

Comment:
When I designed the interface to represent navigation I tried to optimize for a variety of use case:

flat list of pages

```
- page 1
- page 2
- page 3
```
=>
```tsx
[
  { name: 'page-1', title: 'Page 1' },
  { name: 'page-2', title: 'Page 2' },
  { name: 'page-3', title: 'Page 3' }
]
```
flat list with sub headers

```
header 1
------
- page 1
- page 2
header 2
------
- page 3
```
=>
```tsx
[
  { kind: 'header', title: 'Header 1' },
  { name: 'page-1', title: 'Page 1' },
  { name: 'page-2', title: 'Page 2' },
  { kind: 'header', title: 'Header 2' },
  { name: 'page-3', title: 'Page 3' }
]
```
deeply nested

```
- page 1
- folder 1
  - page 2
  - folder 2
    - folder 3
      - page 3
- folder 3
  - page 4
```
=>
```tsx
[
  { name: 'page-1', title: 'Page 1' },
  { 
    name: 'folder-1', 
    title: 'Folder 1',
    children: [
      { name: 'page-2', title: 'Page 2' },
      {
        name: 'folder-2',
        title: 'Folder 2',
        children: [
          { 
            name: 'folder-3', 
            title: 'Folder 3',
            children: [
              { name: 'page-3', title: 'Page 3' },
            ]
          }
        ]
      }
    ]
  },
  { 
    name: 'folder-3', 
    title: 'Folder 3',
    children: [
      { name: 'page-4', title: 'Page 4' }
    ]
  }
]
```

and any combination of pages, folders, and subheaders.

I believe the interface I came up with is the simplest structure that supports all these use cases. Can you give examples for each of those with the interfaces you've designed?

Reply:
The interface I came up with supports multiple lists with different titles too, and up to one level of nesting. 
It would be easy to make it infinitely nestable, but a second level of nesting already starts looking a bit cramped in the sidebar.

Example with 2 lists and an item with nested items (I think this covers all the cases?):
```
const NAVIGATION: Navigation = [
  {
    title: 'Primary items',
    routes: [
      {
        label: 'Dashboard1',
        path: '/dashboard1',
        icon: <DashboardIcon />,
      },
      {
        label: 'Orders1',
        path: '/dashboard/orders1',
        icon: <ShoppingCartIcon />,
        items: [
          {
            label: 'Nested Orders 1',
            path: '/dashboard/orders1/nested1',
            icon: <ShoppingCartIcon />,
          },
          {
            label: 'Nested Orders 2',
            path: '/dashboard/orders1/nested2',
            icon: <ShoppingCartIcon />,
          },
        ]
      },
    ],
  },
  {
    title: 'Secondary items',
    items: [
      {
        label: 'Dashboard 2',
        path: '/dashboard2',
        icon: <DashboardIcon />,
      },
      {
        label: 'Orders2',
        path: '/dashboard/orders2',
        icon: <ShoppingCartIcon />,
      },
    ],
  },
];
```

I can make the titles optional so that not every list needs a title.
I guess it covers the same cases as your suggestion? I find my format easier to read but i don't mind changing it.
One con of your suggestion would maybe be that the relation between a header and its items is not so clear (it's determined by the order in the array).

Reply:
> I think this covers all the cases?

How does this format support a flat list of pages without sections or subfolders?

> One con of your suggestion would maybe be that the relation between a header and its items is not so clear (it's determined by the order in the array).

The relation is similar to the relation between titles in html and paragraphs. It's not a hierarchical relationship, the headers are to be seen as markup. The flat list keeps it extensible for adding more markup in the future

Reply:
> How does this format support a flat list of pages without sections or subfolders?

If we don't want a title I could titles optional and it would be:

```
const NAVIGATION: Navigation = [
  {
    items: [
      {
        label: 'Dashboard1',
        path: '/dashboard1',
        icon: <DashboardIcon />,
      },
      {
        label: 'Orders1',
        path: '/dashboard/orders1',
        icon: <ShoppingCartIcon />,
      },
    ],
  },
];
```

So yeah there's the `items` property that feels a bit extra in this case, could be seen as a con.

> The relation is similar to the relation between titles in html and paragraphs. It's not a hierarchical relationship, the headers are to be seen as markup. The flat list keeps it extensible for adding more markup in the future. (e.g. look at the MUI docs, one could imagine the sidebar ads are another type of markup)

True, it's more extensible, didn't consider that. I don't mind going with it, shouldn't even be a big change!

Reply:
> If we don't want a title I could make title optional so it would be:

I'm not a fan of having the "title" have both the function of overwriting the name, as well as turn on/off certain features.
How do you differentiate between a folder and a subheader then? One being collapsible and the other separating groups of pages on the same level?


🤔 A horizontal separator is probably a better example for another markup element than the ad

Reply:
The title does not turn on/off features, it's just the heading for each top-level list of navigation items and shows up above them.
Collapsible items are navigation items with the `items` property themselves.
The dividers show up automatically between navigation lists.
I added a screenshot to the PR where this shows.

In the header navigation I assume we could use the same system but no titles.

Anyway let's go with your approach.

Reply:
> Collapsible items are navigation items with the items property themselves.

I'm still not fully getting it. How do you differentiate between a section and a folder? Aren't both of them navigation items with the `items` property?

e.g. subheader nested in a folder, or vice-versa:

```
- page1
subheader1
- page2
-- divider --
- page3
- folder1
  - page4
  - subheader2
  - folder2
    - page5
    -- divider --
    - page6
  - page7
```

> The dividers show up automatically between navigation lists.

Ok, but that doesn't allow for dividers between any two arbitrary items.

Reply:
> I'm still not fully getting it. How do you differentiate between a section and a folder? Aren't both of them navigation items with the items property?

In my first approach it's only possible to add headings at the top-level (they're just the `title`), so not inside collapsible items, so you can't have sections inside folders...
So I guess that your approach in more versatile in that way. I'm trying it out!

Reply:
Oh I see, your top-level is different from nested levels

Reply:
I've also just realized that using your suggested approach for subheaders can be tricky with the MUI List component https://mui.com/material-ui/react-list/ since it takes a `subheader` as a prop for the list.
We can probably use subheaders without the prop, we'll just lose the automatic sticky headers.

Reply:
> since it takes a subheader as a prop for the list.

The content of this prop has to be wrapped in a `<ListSubHeader>` component, and these can also be used inside of the list, just like `<Divider>`. See https://mui.com/material-ui/react-list/#sticky-subheader.

Not 100% sure what would be the best approach. We could declare a navigation item with a title and children but no declared path segment as a section. OTOH a flat list could be more versatile towards the future.